controller_interface/src/controller_interface.cpp

```diff
@@ -20,148 +20,62 @@
-        hardware_interface::lifecycle_state_names::INACTIVE);
       break;
     case LifecycleNodeInterface::CallbackReturn::ERROR:
-      on_error(lifecycle_state_);
-      lifecycle_state_ = rclcpp_lifecycle::State(
-        lifecycle_msgs::msg::State::PRIMARY_STATE_FINALIZED,
-        hardware_interface::lifecycle_state_names::FINALIZED);
-      break;
     case LifecycleNodeInterface::CallbackReturn::FAILURE:
+      return return_type::ERROR;
       break;
   }
-  return lifecycle_state_;
-}
-const rclcpp_lifecycle::State & ControllerInterface::activate()
-{
-  if (lifecycle_state_.id() == lifecycle_msgs::msg::State::PRIMARY_STATE_INACTIVE)
-  {
-    switch (on_activate(lifecycle_state_))
-    {
-      case LifecycleNodeInterface::CallbackReturn::SUCCESS:
-        lifecycle_state_ = rclcpp_lifecycle::State(
-          lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE,
-          hardware_interface::lifecycle_state_names::ACTIVE);
-        break;
-      case LifecycleNodeInterface::CallbackReturn::ERROR:
-        on_error(lifecycle_state_);
-        lifecycle_state_ = rclcpp_lifecycle::State(
-          lifecycle_msgs::msg::State::PRIMARY_STATE_FINALIZED,
-          hardware_interface::lifecycle_state_names::FINALIZED);
-        break;
-      case LifecycleNodeInterface::CallbackReturn::FAILURE:
-        break;
-    }
-  }
-  return lifecycle_state_;
+
+  node_->register_on_configure(
+    std::bind(&ControllerInterface::on_configure, this, std::placeholders::_1));
+
+  node_->register_on_cleanup(
+    std::bind(&ControllerInterface::on_cleanup, this, std::placeholders::_1));
+
+  node_->register_on_activate(
+    std::bind(&ControllerInterface::on_activate, this, std::placeholders::_1));
+
+  node_->register_on_deactivate(
+    std::bind(&ControllerInterface::on_deactivate, this, std::placeholders::_1));
+
+  node_->register_on_shutdown(
+    std::bind(&ControllerInterface::on_shutdown, this, std::placeholders::_1));
+
+  node_->register_on_error(std::bind(&ControllerInterface::on_error, this, std::placeholders::_1));
+
+  return return_type::OK;
 }
 
-const rclcpp_lifecycle::State & ControllerInterface::shutdown()
+const rclcpp_lifecycle::State & ControllerInterface::configure()
 {
-  switch (on_shutdown(lifecycle_state_))
-  {
-    case LifecycleNodeInterface::CallbackReturn::SUCCESS:
-      lifecycle_state_ = rclcpp_lifecycle::State(
-        lifecycle_msgs::msg::State::PRIMARY_STATE_FINALIZED,
-        hardware_interface::lifecycle_state_names::FINALIZED);
-      break;
-    case LifecycleNodeInterface::CallbackReturn::ERROR:
-      on_error(lifecycle_state_);
-      lifecycle_state_ = rclcpp_lifecycle::State(
-        lifecycle_msgs::msg::State::PRIMARY_STATE_FINALIZED,
-        hardware_interface::lifecycle_state_names::FINALIZED);
-      break;
-    case LifecycleNodeInterface::CallbackReturn::FAILURE:
-      break;
-  }
-  return lifecycle_state_;
-}
+  update_rate_ = node_->get_parameter("update_rate").as_int();
```

Comment:
This is the only tricky part in this PR. We have to (or at least should add this) because it enables us to read  "update_rate" parameter in configure method and not in init. This also enables easier reset of this parameter without a need to reload the controller. Which could be tricky for setting parameter.